File: src/main/java/studio/awel/FancyCasinos/slots/SlotGUI.java
package studio.awel.FancyCasinos.slots;

import com.samjakob.spigui.buttons.SGButton;
import com.samjakob.spigui.item.ItemBuilder;
import com.samjakob.spigui.menu.SGMenu;
import org.bukkit.Bukkit;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.inventory.ItemFlag;
import org.bukkit.scheduler.BukkitRunnable;
import studio.awel.FancyCasinos.FancyCasinos;
import studio.awel.FancyCasinos.config.ConfigManager;
import studio.awel.FancyCasinos.utilities.ColorFormater;

import java.util.HashMap;
import java.util.Map;
import java.util.Random;

import static studio.awel.FancyCasinos.FancyCasinos.spiGUI;

public class SlotGUI {
    private final Machine machine = new Machine();
    private final ConfigManager configManager;
    private final FancyCasinos plugin;
    private final Random random = new Random();

    private final int[] slotPositions = {
            // Top row
            10, 11, 12,
            // Middle row (winning line)
            19, 20, 21,
            // Bottom row
            28, 29, 30
    };

    private final Map<Material, Double> slotItems = new HashMap<>();

    public SlotGUI(ConfigManager configManager, FancyCasinos plugin) {
        this.configManager = configManager;
        this.plugin = plugin;

        slotItems.put(Material.GOLD_NUGGET, 0.5);
        slotItems.put(Material.GOLD_INGOT, 1.0);
        slotItems.put(Material.RAW_GOLD, 1.5);
        slotItems.put(Material.RAW_GOLD_BLOCK, 2.5);
        slotItems.put(Material.GOLD_BLOCK, 5.0);
    }

    public void openGUI(Player player, double betAmount) {
        SGMenu menu = spiGUI.create(
                ColorFormater.c(configManager.getConfig().slotsTitle()), 5);

        setupSlotMachine(menu);
        player.openInventory(menu.getInventory());
        startSlotAnimation(player, menu, betAmount);
    }

    private void setupSlotMachine(SGMenu menu) {
        // Use minimal background elements
        for (int i = 0; i < 45; i++) {
            // Only set borders with glass
            if (i < 9 || i > 35 || i % 9 == 0 || i % 9 == 8) {
                ItemBuilder glassItem = new ItemBuilder(Material.LIGHT_BLUE_STAINED_GLASS_PANE)
                        .name(" ")
                        .flag(ItemFlag.HIDE_ATTRIBUTES);
                menu.setButton(i, new SGButton(glassItem.build()));
            }
        }

        // Yellow line for winning row
        for (int i = 18; i <= 22; i++) {
            ItemBuilder yellowGlass = new ItemBuilder(Material.YELLOW_STAINED_GLASS_PANE)
                    .name(" ")
                    .flag(ItemFlag.HIDE_ATTRIBUTES);
            menu.setButton(i, new SGButton(yellowGlass.build()));
        }

        // Empty slots
        for (int pos : slotPositions) {
            ItemBuilder defaultItem = new ItemBuilder(Material.WHITE_STAINED_GLASS_PANE)
                    .name(" ")
                    .flag(ItemFlag.HIDE_ATTRIBUTES);
            menu.setButton(pos, new SGButton(defaultItem.build()));
        }

        // Instructions
        ItemBuilder instructionsItem = new ItemBuilder(Material.OAK_SIGN)
                .name(ColorFormater.c(configManager.getConfig().slotsInstructionsTitle()))
                .lore(ColorFormater.c(configManager.getConfig().slotsInstructionsLore()).split("</nl>"))
                .flag(ItemFlag.HIDE_ATTRIBUTES);
        menu.setButton(15, new SGButton(instructionsItem.build()));
    }

    private void startSlotAnimation(Player player, SGMenu menu, double betAmount) {
        int[] animationCount = {0};
        final int maxAnimations = 20;

        new BukkitRunnable() {
            @Override
            public void run() {
                if (animationCount[0] >= maxAnimations) {
                    cancel();

                    Machine.SpinResult winnings = machine.spin(betAmount);
                    displayFinalResult(player, menu, winnings, betAmount);
                    return;
                }

                for (int pos : slotPositions) {
                    Material[] materials = slotItems.keySet().toArray(new Material[0]);
                    Material randomMaterial = materials[random.nextInt(materials.length)];
                    double multiplier = slotItems.get(randomMaterial);

                    ItemBuilder slotItem = new ItemBuilder(randomMaterial)
                            .name(ColorFormater.c("&e&l" + materialToName(randomMaterial).toUpperCase() + " &8(" + multiplier + "x)"))
                            .flag(ItemFlag.HIDE_ATTRIBUTES);

                    menu.setButton(pos, new SGButton(slotItem.build()));
                    player.getOpenInventory().getTopInventory().setItem(pos, slotItem.build());
                }

                player.updateInventory();
                animationCount[0]++;
            }
        }.runTaskTimer(plugin, 5L, 5L);
    }

    private void displayFinalResult(Player player, SGMenu menu, Machine.SpinResult winnings, double betAmount) {
        java.util.List<SlotBlock> rolledBlocks = winnings.getRolledBlocks();

        for (int i = 0; i < rolledBlocks.size(); i++) {
            if (i >= 3) break;

            SlotBlock slotBlock = rolledBlocks.get(i);
            Material material = nameToMaterial(slotBlock.getName());

            ItemBuilder slotItem = new ItemBuilder(material)
                    .name(ColorFormater.c("&f&l" + slotBlock.getName()))
                    .lore(ColorFormater.c("&eMultiplier: &f" + slotBlock.getMultiplier() + "x"))
                    .flag(ItemFlag.HIDE_ATTRIBUTES);

            int pos = 19 + i;
            menu.setButton(pos, new SGButton(slotItem.build()));
            player.getOpenInventory().getTopInventory().setItem(pos, slotItem.build());
        }

        player.updateInventory();

        Bukkit.getScheduler().runTaskLater(plugin, () -> {
            showEndingScreen(player, winnings, betAmount);
        }, 30L);
    }

    private void showEndingScreen(Player player, Machine.SpinResult winnings, double betAmount) {
        boolean isWin = winnings.getFinalAmount() > betAmount;
        String title = isWin ?
                configManager.getConfig().slotsWinTitle() :
                configManager.getConfig().slotsLoseTitle();

        SGMenu resultScreen = spiGUI.create(ColorFormater.c(title), 5);

        // Set minimal styled background (not all green)
        for (int i = 0; i < 45; i++) {
            // Only set borders and keep the original style
            if (i < 9 || i > 35 || i % 9 == 0 || i % 9 == 8) {
                Material borderMaterial = isWin ?
                        Material.LIME_STAINED_GLASS_PANE :
                        Material.RED_STAINED_GLASS_PANE;

                ItemBuilder borderItem = new ItemBuilder(borderMaterial)
                        .name(" ")
                        .flag(ItemFlag.HIDE_ATTRIBUTES);
                resultScreen.setButton(i, new SGButton(borderItem.build()));
            }
        }

        // Display result
        double amountDiff = isWin ? winnings.getFinalAmount() : (betAmount - winnings.getFinalAmount());
        Material resultMaterial = isWin ? Material.EMERALD : Material.BARRIER;

        String resultText = isWin ?
                configManager.getConfig().slotsWinAmount().replace("{amount}", String.format("%.2f", amountDiff)) :
                configManager.getConfig().slotsLoseAmount().replace("{amount}", String.format("%.2f", amountDiff));

        String[] loreText = isWin ?
                configManager.getConfig().slotsWinLore().split("</nl>") :
                configManager.getConfig().slotsLoseLore().split("</nl>");

        for (int i = 0; i < loreText.length; i++) {
            loreText[i] = loreText[i]
                    .replace("{bet}", String.format("%.2f", betAmount))
                    .replace("{final}", String.format("%.2f", winnings.getFinalAmount()));
        }

        ItemBuilder resultItem = new ItemBuilder(resultMaterial)
                .name(ColorFormater.c(resultText))
                .lore(ColorFormater.c(String.join("</nl>", loreText)).split("</nl>"))
                .flag(ItemFlag.HIDE_ATTRIBUTES);

        resultScreen.setButton(22, new SGButton(resultItem.build()));

        // Show rolled blocks
        for (int i = 0; i < winnings.getRolledBlocks().size(); i++) {
            if (i >= 3) break;

            SlotBlock slotBlock = winnings.getRolledBlocks().get(i);
            Material material = nameToMaterial(slotBlock.getName());

            ItemBuilder slotItem = new ItemBuilder(material)
                    .name(ColorFormater.c("&e&l" + slotBlock.getName()))
                    .lore("", ColorFormater.c("&fMultiplier: &e" + slotBlock.getMultiplier() + "x"))
                    .flag(ItemFlag.HIDE_ATTRIBUTES);

            resultScreen.setButton(11 + i, new SGButton(slotItem.build()));
        }

        // Total multiplier
        double multiplier = 1.0;
        for (SlotBlock block : winnings.getRolledBlocks()) {
            multiplier *= block.getMultiplier();
        }

        ItemBuilder multiplierItem = new ItemBuilder(Material.EXPERIENCE_BOTTLE)
                .name(ColorFormater.c(configManager.getConfig().slotsTotalMultiplierTitle()))
                .lore("", ColorFormater.c(configManager.getConfig().slotsTotalMultiplierValue()
                        .replace("{multiplier}", String.format("%.2f", multiplier))))
                .flag(ItemFlag.HIDE_ATTRIBUTES);

        resultScreen.setButton(31, new SGButton(multiplierItem.build()));

        // Play again button
        ItemBuilder playAgainItem = new ItemBuilder(Material.EMERALD)
                .name(ColorFormater.c(configManager.getConfig().slotsPlayAgainTitle()))
                .flag(ItemFlag.HIDE_ATTRIBUTES);

        SGButton playAgainButton = new SGButton(playAgainItem.build()).withListener(event -> {
            Player user = (Player) event.getWhoClicked();
            user.closeInventory();
            // Return to main menu
            plugin.getCommand("casino").execute(user, "casino", new String[0]);
        });

        resultScreen.setButton(38, playAgainButton);

        // Exit button
        ItemBuilder exitItem = new ItemBuilder(Material.BARRIER)
                .name(ColorFormater.c(configManager.getConfig().slotsExitTitle()))
                .flag(ItemFlag.HIDE_ATTRIBUTES);

        SGButton exitButton = new SGButton(exitItem.build()).withListener(event -> {
            ((Player) event.getWhoClicked()).closeInventory();
        });

        resultScreen.setButton(42, exitButton);

        player.openInventory(resultScreen.getInventory());
    }

    private Material nameToMaterial(String blockName) {
        if (blockName.toLowerCase().contains("gold nugget")) {
            return Material.GOLD_NUGGET;
        } else if (blockName.toLowerCase().contains("gold ingot")) {
            return Material.GOLD_INGOT;
        } else if (blockName.toLowerCase().contains("raw gold") && blockName.toLowerCase().contains("block")) {
            return Material.RAW_GOLD_BLOCK;
        } else if (blockName.toLowerCase().contains("raw gold")) {
            return Material.RAW_GOLD;
        } else if (blockName.toLowerCase().contains("gold block")) {
            return Material.GOLD_BLOCK;
        } else {
            return Material.BARRIER;
        }
    }

    private String materialToName(Material material) {
        switch (material) {
            case GOLD_NUGGET: return "Gold Nugget";
            case GOLD_INGOT: return "Gold Ingot";
            case RAW_GOLD: return "Raw Gold";
            case RAW_GOLD_BLOCK: return "Raw Gold Block";
            case GOLD_BLOCK: return "Gold Block";
            default: return "Unknown";
        }
    }
}

File: src/main/java/studio/awel/FancyCasinos/slots/SlotGUI.kt
package studio.awel.FancyCasinos.slots

import com.samjakob.spigui.buttons.SGButton
import com.samjakob.spigui.item.ItemBuilder
import com.samjakob.spigui.menu.SGMenu
import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.inventory.ItemFlag
import org.bukkit.scheduler.BukkitRunnable
import studio.awel.FancyCasinos.FancyCasinos
import studio.awel.FancyCasinos.utilities.MoneyUtil

import studio.awel.FancyCasinos.utilities.cigan.translate
import java.util.function.Consumer

class SlotGUI {

    val machine = Machine()

    private val slotPositions = arrayOf(
        // Top row
        10, 11, 12,
        // Middle row (winning line)
        19, 20, 21,
        // Bottom row
        28, 29, 30
    )

    // Updated to use only gold-related items
    private val slotItems = mapOf(
        Material.GOLD_NUGGET to 0.5,
        Material.GOLD_INGOT to 1.0,
        Material.RAW_GOLD to 1.5,
        Material.RAW_GOLD_BLOCK to 2.5,
        Material.GOLD_BLOCK to 5.0
    )

    fun openMenu(player: Player) {
        val menu: SGMenu = FancyCasinos.spiGUI.create("Slots", 5)

        setupSlotMachine(menu)

        val spin = ItemBuilder(Material.CLOCK)
            .name("&a&lSPIN")
            .lore(
                "",
                "&8&l| &aClick here to pay for a spin",
                "&8&l| &aYou are betting &a&n$0&a."
            )
            .flag(ItemFlag.HIDE_ATTRIBUTES)
            .build()

        val spinButton = SGButton(spin).withListener { event: InventoryClickEvent ->
            val user = event.whoClicked as Player
            user.closeInventory()

            MoneyUtil.typeInChat(user, "Slots Bet", 15L,
                Consumer<String> { input ->
                    val amount = input.toDoubleOrNull()
                    if (amount == null) {
                        user.sendMessage("&cInvalid amount. Please enter a number.".translate())
                        return@Consumer
                    } else {
                        user.sendMessage("&aYou have bet $amount.".translate())

                        val animationMenu = FancyCasinos.spiGUI.create("Slots", 5)
                        setupSlotMachine(animationMenu)
                        user.openInventory(animationMenu.inventory)

                        startSlotAnimation(user, animationMenu, amount)
                    }
                },
                Runnable {
                    user.sendMessage("&cBetting cancelled.".translate())
                }
            )
        }

        menu.setButton(24, spinButton)
        player.openInventory(menu.inventory)
    }

    private fun setupSlotMachine(menu: SGMenu) {
        for (i in 0 until 45) {
            val glassItem = ItemBuilder(Material.GRAY_STAINED_GLASS_PANE)
                .name(" ")
                .flag(ItemFlag.HIDE_ATTRIBUTES)
                .build()

            menu.setButton(i, SGButton(glassItem))
        }

        for (i in 18..22) {
            val yellowGlass = ItemBuilder(Material.YELLOW_STAINED_GLASS_PANE)
                .name("")
                .flag(ItemFlag.HIDE_ATTRIBUTES)
                .build()

            menu.setButton(i, SGButton(yellowGlass))
        }

        for (pos in slotPositions) {
            val defaultItem = ItemBuilder(Material.WHITE_STAINED_GLASS_PANE)
                .name(" ")
                .flag(ItemFlag.HIDE_ATTRIBUTES)
                .build()

            menu.setButton(pos, SGButton(defaultItem))
        }

        val instructionsItem = ItemBuilder(Material.OAK_SIGN)
            .name("&7&lINSTRUCTIONS")
            .lore(
                "",
                "&8&l| &7You pay for spins and every spin makes a",
                "&8&l| &7random amount of money. You can get double",
                "&8&l| &7the money if you get a golden",
                "&8&l| &7spin."
            )
            .flag(ItemFlag.HIDE_ATTRIBUTES)
            .build()
        menu.setButton(15, SGButton(instructionsItem))

        val earningsItem = ItemBuilder(Material.CHEST)
            .name("&f&lEARNINGS")
            .lore(
                "",
                "&8&l| &fYou have earned &a&n$0&f in total",
                "&8&l| &fYou have earned &a&n$0&f in this spin.",
                "&7",
                "&8&l| &fClick to claim &f&n$0&f.",
            )
            .flag(ItemFlag.HIDE_ATTRIBUTES)
            .build()
        menu.setButton(33, SGButton(earningsItem))
    }

    private fun startSlotAnimation(player: Player, menu: SGMenu, betAmount: Double) {
        var animationCount = 0
        val maxAnimations = 20

        object : BukkitRunnable() {
            override fun run() {
                if (animationCount >= maxAnimations) {
                    cancel()

                    val winnings = machine.spin(betAmount.toDouble())
                    displayFinalResult(player, menu, winnings, betAmount)
                    return
                }

                for (pos in slotPositions) {
                    val randomMaterial = slotItems.keys.random()
                    val multiplier = slotItems[randomMaterial] ?: 1.0

                    val slotItem = ItemBuilder(randomMaterial)
                        .name("&e&l${materialToName(randomMaterial).toUpperCase()} &8(${multiplier}x)")
                        .flag(ItemFlag.HIDE_ATTRIBUTES)
                        .build()

                    menu.setButton(pos, SGButton(slotItem))
                    player.openInventory.topInventory.setItem(pos, slotItem)
                }

                player.updateInventory()
                animationCount++
            }
        }.runTaskTimer(FancyCasinos.getPlugin(FancyCasinos::class.java), 5L, 5L)
    }

    private fun displayFinalResult(player: Player, menu: SGMenu, winnings: Machine.SpinResult, betAmount: Double) {
        val rolledBlocks = winnings.rolledBlocks

        for (i in rolledBlocks.indices) {
            if (i >= 3) break

            val slotBlock = rolledBlocks[i]
            val material = nameToMaterial(slotBlock.name)

            val slotItem = ItemBuilder(material)
                .name("&f&l${slotBlock.name}")
                .lore("&eMultiplier: &f${slotBlock.multiplier}x")
                .flag(ItemFlag.HIDE_ATTRIBUTES)
                .build()

            val pos = 19 + i
            menu.setButton(pos, SGButton(slotItem))
            player.openInventory.topInventory.setItem(pos, slotItem)
        }

        player.updateInventory()

        Bukkit.getScheduler().runTaskLater(
            FancyCasinos.getPlugin(
                FancyCasinos::class.java), Runnable {
            showEndingScreen(player, winnings, betAmount)
        }, 30L)
    }

    private fun showEndingScreen(player: Player, winnings: Machine.SpinResult, betAmount: Double) {
        val isWin = winnings.finalAmount > betAmount
        val title = if (isWin) "&a&lYOU WON!" else "&c&lYOU LOST!"
        val resultScreen = FancyCasinos.spiGUI.create(title.translate(), 5)

        val bgMaterial = if (isWin) Material.LIME_STAINED_GLASS_PANE else Material.RED_STAINED_GLASS_PANE
        for (i in 0 until 45) {
            val bgItem = ItemBuilder(bgMaterial)
                .name(" ")
                .flag(ItemFlag.HIDE_ATTRIBUTES)
                .build()
            resultScreen.setButton(i, SGButton(bgItem))
        }

        val amountDiff = if (isWin) winnings.finalAmount else (betAmount - winnings.finalAmount)
        val resultMaterial = if (isWin) Material.EMERALD else Material.BARRIER
        val resultItem = ItemBuilder(resultMaterial)
            .name(if (isWin) "&a&l+$${amountDiff}" else "&c&l-$${amountDiff}")
            .lore(
                "",
                "&7You bet: &a&n$${betAmount}",
                "&7Final amount: &a&n$${winnings.finalAmount}",
                "",
                if (isWin) "&aWell Done!" else "&cBetter luck next time!"
            )
            .flag(ItemFlag.HIDE_ATTRIBUTES)
            .build()
        resultScreen.setButton(22, SGButton(resultItem))

        for (i in winnings.rolledBlocks.indices) {
            if (i >= 3) break

            val slotBlock = winnings.rolledBlocks[i]
            val material = nameToMaterial(slotBlock.name)

            val slotItem = ItemBuilder(material)
                .name("&e&l${slotBlock.name}")
                .lore(
                    "",
                    "&fMultiplier: &e${slotBlock.multiplier}x"
                )
                .flag(ItemFlag.HIDE_ATTRIBUTES)
                .build()

            resultScreen.setButton(11 + i, SGButton(slotItem))
        }

        val multiplier = winnings.rolledBlocks.fold(1.0) { acc, block -> acc * block.multiplier }
        val multiplierItem = ItemBuilder(Material.EXPERIENCE_BOTTLE)
            .name("&e&lTotal Multiplier")
            .lore(
                "",
                "&fx${String.format("%.2f", multiplier)}"
            )
            .flag(ItemFlag.HIDE_ATTRIBUTES)
            .build()
        resultScreen.setButton(31, SGButton(multiplierItem))

        val playAgainItem = ItemBuilder(Material.EMERALD)
            .name("&a&lPLAY AGAIN")
            .flag(ItemFlag.HIDE_ATTRIBUTES)
            .build()
        val playAgainButton = SGButton(playAgainItem).withListener { event: InventoryClickEvent ->
            val user = event.whoClicked as Player
            user.closeInventory()

            MoneyUtil.typeInChat(user, "Slots Bet", 15L,
                Consumer<String> { input ->
                    val amount = input.toDoubleOrNull()
                    if (amount == null) {
                        user.sendMessage("&cInvalid amount. Please enter a number.".translate())
                        return@Consumer
                    } else {
                        user.sendMessage("&aYou have bet $amount.".translate())

                        val animationMenu = FancyCasinos.spiGUI.create("Slots", 5)
                        setupSlotMachine(animationMenu)
                        user.openInventory(animationMenu.inventory)

                        startSlotAnimation(user, animationMenu, amount)
                    }
                },
                Runnable {
                    user.sendMessage("&cBetting cancelled.".translate())
                }
            )
        }
        resultScreen.setButton(38, playAgainButton)

        val exitItem = ItemBuilder(Material.BARRIER)
            .name("&c&lEXIT")
            .flag(ItemFlag.HIDE_ATTRIBUTES)
            .build()
        val exitButton = SGButton(exitItem).withListener { event: InventoryClickEvent ->
            (event.whoClicked as Player).closeInventory()
        }
        resultScreen.setButton(42, exitButton)

        player.openInventory(resultScreen.inventory)
    }

    private fun nameToMaterial(blockName: String): Material {
        return when {
            blockName.contains("Gold Nugget", ignoreCase = true) -> Material.GOLD_NUGGET
            blockName.contains("Gold Ingot", ignoreCase = true) -> Material.GOLD_INGOT
            blockName.contains("Raw Gold", ignoreCase = true) && blockName.contains("Block", ignoreCase = true) -> Material.RAW_GOLD_BLOCK
            blockName.contains("Raw Gold", ignoreCase = true) -> Material.RAW_GOLD
            blockName.contains("Gold Block", ignoreCase = true) -> Material.GOLD_BLOCK
            else -> Material.BARRIER
        }
    }

    private fun materialToName(material: Material): String {
        return when (material) {
            Material.GOLD_NUGGET -> "Gold Nugget"
            Material.GOLD_INGOT -> "Gold Ingot"
            Material.RAW_GOLD -> "Raw Gold"
            Material.RAW_GOLD_BLOCK -> "Raw Gold Block"
            Material.GOLD_BLOCK -> "Gold Block"
            else -> "Unknown"
        }
    }
}